src/components/control/select.jsx

```diff
@@ -0,0 +1,295 @@
+
+  const classNames = {
+    input: invalid ? styles['input-invalid'] : styles.input,
+  };
+
+  const ids = {
+    input: props.id,
+    label: props.labelledBy,
+    listbox: `${props.id}-single-choice-listbox`,
+    tooltip: `${props.id}-tooltip`,
+    validation: `${props.id}Hint`,
+  };
+
+  const clear = () => {
+    setValue(null);
+    setSearchValue(undefined);
+    refs.input.current.focus();
+  };
+
+  function onClick() {
+    if (!props.readOnly) setShowOptions(true);
+  }
+
+  function onKeyDown(event) {
+    switch (event.key) {
+      case 'Enter':
+        event.preventDefault();
+        if (!showOptions && !props.readOnly) setShowOptions(true);
+        break;
+      case 'Escape':
+        event.preventDefault();
+        setShowOptions(false);
+        break;
+      default:
+    }
+  }
+
+  function onSelectionChange(event) {
+    const selectedValue = event.target.value;
+    setValue(selectedValue);
+    setSearchValue(selectedValue.label);
+    setShowOptions(false);
+    refs.input.current.focus();
+  }
+
+  function onSearchChange(event) {
+    const newValue = event.target.value;
+
+    if (newValue === '') {
+      clear();
+      return;
+    }
+
+    setSearchValue(newValue);
+    setShowOptions(true);
+  }
+
+  useImperativeHandle(selfRef, () => ({
+    get dirty() {
+      return props.value === undefined ? value !== null : props.value !== value;
+    },
+    get value() {
+      return value === null ? undefined : value;
+    },
+  }));
+
+  useEffect(() => {
+    props.onInvalid({
+      target: selfRef.current,
+      detail: {
+        validationMessage,
+      },
+    });
+  }, [validationMessage]);
+
+  useEffect(() => {
+    if (!mounted.current) return;
+
+    if (props.value === undefined) return;
+
+    setValue(props.value);
+  }, [props.value]);
+
+  useEffect(() => {
+    if (!mounted.current) return;
+
+    if (hasBeenBlurred) {
+      validate();
+    }
+
+    if (props.value === value) return;
+
+    props.onChange({ target: selfRef.current });
+  }, [value]);
+
+  function handleBlur(event) {
+    if (!(refs.listbox.current && refs.listbox.current.contains(event.relatedTarget))) {
+      validate();
+    }
+
+    setBeenBlurred(true);
+  }
+
+  useEffect(function updateOptionVisibility() {
+    if (searchValue) {
+      props.listOptionRefs.forEach((listOptionRef) => {
+        if (listOptionRef.current) {
+          if (props.displayValueEvaluator) {
+            const listOptionDisplayValue = props.displayValueEvaluator(listOptionRef.current.value);
+
+            const matches = listOptionDisplayValue.toLowerCase().includes(searchValue.toLowerCase());
+            listOptionRef.current.setVisible(matches);
+          }
+        }
+      });
+    } else {
+      props.listOptionRefs.forEach((listOptionRef) => {
+        if (listOptionRef.current) {
+          listOptionRef.current.setVisible(true);
+        }
+      });
+    }
+  }, [searchValue, props.listOptionRefs]);
```

Comment:
Since this is in a `forEach`, I think we could get rid of the `ifs` entirely and use `filter`. 

There is also some restructuring we could do to this entire `useEffect` to reduce some of the redundancy and nesting. Both the `if` and `else` only care about `props.listOptionRefs` where there is a `current`, so we could hoist that out of the nested logic.

I also made an early `return` for the `displayValueEvaluator` falsiness case to preserve existing logic.

```javascript
useEffect(() => {
  const listOptionRefs = props.listOptionRefs.filter(r => r.current);
  if(searchValue) {
    if(!displayValueEvaluator) return;
    
    listOptionRefs.forEach(listOptionRef => {
      const listOptionDisplayValue = props.displayValueEvaluator(listOptionRef.current.value);
      const matches = listOptionDisplayValue.toLowerCase().includes(searchValue.toLowerCase());
      listOptionRef.current.setVisible(matches);
    }
  } else {
    listOptionRefs.forEach(listOptionRef => listOptionRef.current.setVisible(true));
  }
})
```

Still a bit complex but it has less nesting. As usual this is untested so I might have missed something but I think that's green-green.

Reply:
I'll give it a go. Since we're going for a full round of QA on this component, it might be a good time to do this refactor.